File: packages/yarilo/src/vm.ts
//
// Copyright © 2020 Anticrm Platform Contributors.
// 
// Licensed under the Eclipse Public License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License. You may
// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// 
// See the License for the specific language governing permissions and
// limitations under the License.
//

export type Proc = (pc: PC) => any
type Dict = { [key: string]: any }

export type Bound = { 
  get: (sym: string) => any
  set: (sym: string, value: any) => void
}

export enum WordKind { 
  Word = 0,
  GetWord,
  SetWord,
  Quote
}

export class Word {
  readonly kind: WordKind
  readonly sym: string
  bound?: Bound

  constructor(kind: WordKind, sym: string) {
    this.kind = kind
    this.sym = sym
  }
}

export type CodeItem = Word | number | string | CodeItem[]
export type Code = CodeItem[]

export function bind(code: Code, boundFactory: (sym: string) => Bound | undefined) {
  let i = 0
  while (i < code.length) {
    const item = code[i]
    if (typeof item === 'object') {
      const word = item as Word
      if (word.sym) {
        const bound = boundFactory(word.sym)
        if (bound) {
          word.bound = bound
        }
      } else {
        if (item.hasOwnProperty('path')) {
          const path = (item as any).path
          const bound = boundFactory(path[0])
          if (bound) {
            word.bound = bound
          }
        } else {
          const code = item as Code
          bind(code, boundFactory)
        }
      }
    }
    i++
  }
}

export class VM {
  dictionary: Dict = {}
  stack: any[] = []

  bind(code: Code) {
    bind(code, () => {
      return { 
        get: (sym: string) => this.dictionary[sym],
        set: (sym: string, value: any) => this.dictionary[sym] = value
      }
    })
  }

  exec(code: Code): any {
    return new PC(this, code).exec()
  }

}

export class PC { 
  code: Code
  pc: number
  vm: VM

  constructor(vm: VM, code: Code) { 
    this.code = code
    this.pc = 0
    this.vm = vm
  }

  next(): any {
    const item = this.code[this.pc++]
    switch (typeof item) {
      case 'object':
        const word = item as Word
        switch (word.kind) {
          case WordKind.SetWord: 
            const x = this.next()
            if (!word.bound)
              throw new Error('word not bound ' + word)
            word.bound.set(word.sym, x)
            return x
          case undefined:
            if (item.hasOwnProperty('path')) {
              const path = (item as any).path as string[]
              const bound = (item as any).bound
              if (!bound)
                throw new Error('path not bound')
              return path.slice(1).reduce((acc, val) => acc[val], bound.get(path[0]))
            } else {
              return item
            }
          default:
            if (!word.bound)
              throw new Error('word not bound ' + word)
            const f = word.bound.get(word.sym)
            return typeof f === 'function' ? f(this) : f
        }
      case 'number':
      case 'string':
        return item
    }
  }

  exec(): any { 
    let result
    while (this.pc < this.code.length) {
      result = this.next()
    }
    return result
  }
}

export interface Context {
  vm: VM
}

File: yar/vm.go
//
// Copyright © 2020 Anticrm Platform Contributors.
//
// Licensed under the Eclipse Public License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License. You may
// obtain a copy of the License at https://www.eclipse.org/legal/epl-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//
// See the License for the specific language governing permissions and
// limitations under the License.
//

package yar

type ProcFunc = func(vm *VM) Value

type VM struct {
	code   int
	pc     int
	sp     int
	bp     int
	result Value

	stack       []Value
	heap        []Value
	wordBinding *[32]Value

	words   []_Word
	blocks  []_Block
	maps    []_Map
	natives []_Native
	procs   []_Proc
	strings []string

	Dictionary Map

	execVmt [LastType]func(vm *VM, value Value) Value

	symbols        map[string]sym
	InverseSymbols map[sym]string
	lastSymbol     sym

	Library *Library
}

func (vm *VM) alloc(v Value) int {
	result := len(vm.heap)
	vm.heap = append(vm.heap, v)
	return result
}

func NewVM(stackSize int) *VM {
	vm := &VM{
		stack:       make([]Value, stackSize),
		heap:        make([]Value, 0),
		wordBinding: new([32]Value),

		words:  make([]_Word, 0),
		blocks: make([]_Block, 0),
		maps:   make([]_Map, 0),

		symbols:        make(map[string]sym),
		InverseSymbols: make(map[sym]string),
		lastSymbol:     0,
	}

	vm.Dictionary = vm.allocMap()
	vm.Dictionary.put(vm, vm.GetSymbol("load-native"), vm.allocNative(loadNative, "boot/load-native").Value())

	vm.execVmt = execVmt

	vm.Library = &Library{}

	return vm
}

func (vm *VM) push(v Value) {
	vm.stack[vm.sp] = v
	vm.sp++
}

func (vm *VM) GetSymbol(s string) sym {
	sym, ok := vm.symbols[s]
	if !ok {
		vm.lastSymbol++
		sym = vm.lastSymbol
		vm.symbols[s] = sym
		vm.InverseSymbols[sym] = s
	}
	return sym
}

func (vm *VM) Bind(code Block) {
	code.Bind(vm, &BindToHeap{vm: vm, m: vm.Dictionary.Value().Val()})
}

func (vm *VM) BindAndExec(code Block) Value {
	vm.Bind(code)
	return vm.Exec(code)
}

func (vm *VM) nextNoInfix() Value {
	value := vm.heap[vm.blocks[vm.code].values[vm.pc]]
	vm.pc++
	result := value.exec(vm)
	vm.result = result
	return result
}

func (vm *VM) ReadNext() Value {
	value := vm.heap[vm.blocks[vm.code].values[vm.pc]]
	vm.pc++
	vm.result = value
	return value
}

func (vm *VM) Next() Value {
	return vm.nextNoInfix()
}

func (vm *VM) Exec(block Block) Value {
	b := vm.blocks[Value(block).Val()]
	codeSave := vm.code
	pcSave := vm.pc
	vm.code = block.Value().Val()
	vm.pc = 0
	len := len(b.values)
	var result Value
	for vm.pc < len {
		result = vm.Next()
	}
	vm.pc = pcSave
	vm.code = codeSave
	return result
}
